tools/aliveness/src/main.rs

```diff
@@ -0,0 +1,358 @@
+    }
+
+    pub fn new() -> Self {
+        let sender = Self::create_sender();
+        Self { sender }
+    }
+
+    pub async fn read_from_hula(&mut self) -> Result<RobotState> {
+        let (sender, receiver) = oneshot::channel();
+        self.sender.send(sender).await?;
+        Ok(receiver.await?)
+    }
+}
+
+struct RobotInfo {
+    hulks_os_version: String,
+    hostname: String,
+    body_id: Option<String>,
+    head_id: Option<String>,
+    battery: Option<Battery>,
+}
+
+impl RobotInfo {
+    async fn initialize() -> Result<Self> {
+        let hulks_os_version = get_hulks_os_version()
+            .await
+            .wrap_err("failed to load HULKs-OS version")?;
+        let hostname = hostname::get()
+            .wrap_err("failed to query hostname")?
+            .into_string()
+            .map_err(|hostname| eyre!("invalid utf8 in hostname: {hostname:?}"))?;
+
+        Ok(Self {
+            hulks_os_version,
+            hostname,
+            body_id: None,
+            head_id: None,
+            battery: None,
+        })
+    }
+}
+
+struct AlivenessService {
+    token: CancellationToken,
+    handle: JoinHandle<()>,
+}
+
+impl AlivenessService {
+    fn cancel(&self) {
+        self.token.cancel();
+    }
+
+    async fn join(self) {
+        self.handle.await.unwrap();
+    }
+}
+
+async fn listen_for_network_change(interface_name: String) -> Result<()> {
+    let dbus_conn = Connection::system().await?;
+
+    let link_object = get_link_object(&interface_name, &dbus_conn).await?;
+
+    let rule = MatchRule::builder()
+        .msg_type(MessageType::Signal)
+        .sender("org.freedesktop.network1")?
+        .path(link_object)?
+        .interface("org.freedesktop.DBus.Properties")?
+        .member("PropertiesChanged")?
+        .build();
+
+    let mut stream = MessageStream::for_match_rule(rule, &dbus_conn, Some(1)).await?;
+
+    let mut service = None;
+
+    if let Some(ip) = get_ip(&interface_name, &dbus_conn).await? {
+        service = Some(join_multicast(ip, dbus_conn.clone(), interface_name.clone()).await?);
+    }
+
+    while let Some(Ok(msg)) = stream.next().await {
+        if let Ok((_, data, _)) = msg.body::<(String, HashMap<String, Value>, Vec<String>)>() {
+            if let Some(Value::Str(data)) = data.get("IPv4AddressState") {
+                match data.as_str() {
+                    "routable" => {
+                        info!("IPv4 on {} back online", interface_name);
+                        let ip = get_ip(&interface_name, &dbus_conn)
+                            .await?
+                            .ok_or(eyre!("failed to get IP"))?;
+                        service = Some(
+                            join_multicast(ip, dbus_conn.clone(), interface_name.clone()).await?,
+                        );
+                    }
+                    "off" => {
+                        info!("IPv4 on {} offline", interface_name);
+                        if let Some(s) = service {
```

Comment:
Please don't abbreviate 😉 Just shadow the variable inside the if.

Reply:
Shadowing is in this case not possible, since `service` is also accessed within the if. I now renamed `service` to `service_option` and `s` to `service` instead.

Reply:
Yes, you're right. I didn't see that. Looks better now.